skiko/src/jvmMain/objectiveC/macos/Drawlayer.mm

```diff
@@ -281,90 +273,105 @@ - (void) makeFullscreen: (BOOL) value
 
 JNIEXPORT void JNICALL Java_org_jetbrains_skiko_HardwareLayer_nativeInit(JNIEnv *env, jobject canvas, jlong platformInfoPtr)
 {
-    if (layerStorage == nil)
-    {
-        layerStorage = [[NSMutableSet alloc] init];
-    }
+    @autoreleasepool {
+        if (layerStorage == nil)
+        {
+            layerStorage = [[NSMutableSet alloc] init];
+        }
 
-    LayerHandler *layersSet = [[LayerHandler alloc] init];
-    NSObject<JAWT_SurfaceLayers>* dsi_mac = (__bridge NSObject<JAWT_SurfaceLayers> *) platformInfoPtr;
-    layersSet.container = [dsi_mac windowLayer];
-    jobject canvasGlobalRef = env->NewGlobalRef(canvas);
-    [layersSet setCanvasGlobalRef: canvasGlobalRef];
-    layersSet.window = findWindow(platformInfoPtr);
+        LayerHandler *layer = [[LayerHandler alloc] init];
+        NSObject<JAWT_SurfaceLayers>* dsi_mac = (__bridge NSObject<JAWT_SurfaceLayers> *) (void*) platformInfoPtr;
+        layer.container = [dsi_mac windowLayer];
+        jobject canvasGlobalRef = env->NewGlobalRef(canvas);
+        [layer setCanvasGlobalRef: canvasGlobalRef];
+        layer.window = findWindow(platformInfoPtr);
 
-    [layerStorage addObject: layersSet];
+        [layerStorage addObject: layer];
+    }
 }
 
 JNIEXPORT void JNICALL Java_org_jetbrains_skiko_HardwareLayer_nativeDispose(JNIEnv *env, jobject canvas)
 {
-    LayerHandler *layer = findByObject(env, canvas);
-    if (layer != NULL)
-    {
-        [layerStorage removeObject: layer];
-        env->DeleteGlobalRef(layer.canvasGlobalRef);
-        [layer release];
+    @autoreleasepool {
+        LayerHandler *layer = findByObject(env, canvas);
+        if (layer != NULL)
+        {
+            [layerStorage removeObject: layer];
+            env->DeleteGlobalRef(layer.canvasGlobalRef);
+        }
     }
 }
 
 JNIEXPORT jboolean JNICALL Java_org_jetbrains_skiko_PlatformOperationsKt_osxIsFullscreenNative(JNIEnv *env, jobject properties, jobject component)
 {
-    LayerHandler *layer = findByObject(env, component);
-    if (layer != NULL)
-    {
-        return [layer isFullScreen];
+    @autoreleasepool {
```

Comment:
Do we need autorelease pools like that with ARC?

Reply:
Yes, we still need pools. Without them, we have memory leaks.

I just followed the simple rule - if the function is JNI function, it should be wrapped into `autoreleasepool`, without thinking about what this function do, or from where it can be called. It simplifies the code, and reduce the number of possible leaks.

Reply:
It's not exactly like that. If it's a JNI function in .mm file at least. Our Skia wrappers do not need such thing. And I still consider that as a hack.

Reply:
> If it's a JNI function in .mm file at least.

Yes, in mm file, wher we use Objective C style functions.

> And I still consider that as a hack.

Any option we use to fix that will be a hack. The option just to create pool and clear every frame I consider a very huge hack. Can explain, if interested.

Reply:
It really depends on performance. So far, none of our ObjC JNI methods are high frequency, so that would work, I guess.